### DISCUSSION TITLE
WinUI Community Call: June 5, 2024

### DISCUSSION THREAD

**Discussion by duncanmacmichael** [microsoft/microsoft-ui-xaml#9653] [General]
### Please join us for the June 2024 edition of the WinUI Community Call. This event will be live-streamed to YouTube: https://www.youtube.com/watch?v=pExPo1VJ8Ks
#
# Details
Date: Jun 05, 2024
Time: 16:00-17:00 UTC (9:00-10:00am Pacific)

Anyone and everyone is welcome - no pre-registration is required.
This will be an informal interactive live stream directly with members of our engineering team.

## Format
The community call is a call among the WinUI team that is live-streamed onto YouTube. We present on new updates, share information, welcome guests, and answer your questions.

The WinUI and WindowsAppSDK teams are back from a busy //Build season and are excited to share our 1.6 plans and demos! We've also got a few great guests for you on this call. We can't wait to share what's coming next and look forward to seeing you all there!

### Agenda
- Intro
- //Build 2024 recap w/WPF Team guest
- 1.6 experimental overview
- 1.6 //Build TitleBar demo
- 1.6 Tab Tear demo
- Q&A
 
### **Q&A Code of Conduct**
- To help ensure Q&A remains beneficial for the majority of the call's developer audience, please consider whether your question could be framed so that it might be more broadly relevant to our larger developer audience by ensuring it is reasonably concise, broadly beneficial, well-suited for the fast-paced live Q&A format, or able to spark open ended discussion.
- Please understand some questions are not well-suited for the fast-paced live Q&A format and should be filed as standalone [Q&A](https://github.com/microsoft/microsoft-ui-xaml/discussions/new?category=q-a&title=%5BYour+Question+Here%5D) discussions where they can be most appropriately handled. Excluding presentation topic requests, this generally includes requests for dedicated follow-ups or technically complex questions that require investigative time or expertise not included in this month's guest list.
 
### Leave us your questions in the comments on this issue or live in the YouTube chat during the stream!

**Comment by nikolayvpavlov** [comment]
Would someone comment on issue [9154](https://github.com/microsoft/microsoft-ui-xaml/discussions/9154) please?  Performance is important for LoB apps, where we need to display a lot of data items on screen and even WPF was struggling in such scenarios despite virtualization.

**Comment by ReneeGA2020** [comment]
I noticed that winui ICollectionView does not have Filter property, what's the current best practice for filtering a listview/treeview with MVVM then?

I also noticed that I cannot set the hover delay for the tooltip, any way to set the custom delay in may app?

**Reply by AndrewKeepCoding** [reply]
Just so you know, the [AdvancedCollectionView](https://learn.microsoft.com/en-us/windows/communitytoolkit/helpers/advancedcollectionview?wt.mc_id=MVP_356303) from the CommunityToolkit comes with filter feature.

**Reply by ReneeGA2020** [reply]
that’s great to know

**Comment by ReneeGA2020** [comment]
At this point the TreeView's ItemTemplateSelector is basiclly broken because the container recycling puts you into a wrong container - there are lots of bugs issued around that.

The Remark section of the SelectTemplateCore method group says you should use SelectTemplateCore(Object, DependencyObject) instead of SelectTemplateCore(Object) for all VirtualizingStackPanels
https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.datatemplateselector.selecttemplatecore?view=windows-app-sdk-1.5


But this workarounds can only fix ListView and GridView, because by using SelectTemplateCore(Object, DependencyObject), you can no longer set the containers in DataTemplates
 (the Selector will assign a standalone container for you), while for the TreeView,  you will need a TreeViewItem in the DataTemplate, otherwise you will have no way to bind ItemsSource and IsExpaned property for the child levels of the tree.



Does the WinUI team aware of this issue and if so, any plan to address this? Perhaps, adding a [HierarchicalDataTemplate ](https://learn.microsoft.com/en-us/dotnet/api/system.windows.hierarchicaldatatemplate?view=windowsdesktop-8.0) class that allow you to bind data source on the DataTemplate layer like wpf did?

**Comment by pjmlp** [comment]
I missed any updates during the BUILD sessions on actual Visual Studio C++ tooling to bring the developer experience closer to Qt and C++ Builder, or back to what C++/CX used to offer us.

**Comment by AndrewKeepCoding** [comment]
I'd like to know about the coming `TableView` control if there is something you can share at this point:
- Will it have data virtualization and work with large collections? (for example, > 10,000 items)
- Can we switch the rows and columns and add items horizontally?

**Comment by lolametro** [comment]
For 1.6 and control improvements, is https://github.com/microsoft/microsoft-ui-xaml/issues/6945 being looked at? Currently, some (even minor) changes to the default styles require copying and editing the entire XAML style from `generic.xaml` which is a pain. 

As a first step, easy to implement step, it would be nice if all `StaticResource` bindings within control styles would be replaced with `ThemeResource`, so at least lightweight styling (overriding the resources) could be used. IMO this is a low hanging fruit, as it is mainly only search and replace 😊

**Comment by w-ahmad** [comment]
Question: Is there any other option to bring a specific item into view in a `ListView`?

When dealing with a `ListView` containing duplicate items in its `ItemsSource`, there isn't a direct method to specifically bring a duplicate item into view. For instance, if an item appears at index 30 and again at index 100, the challenge is how to scroll to the item at index 100. The commonly known method `ListViewBase.ScrollIntoView(object item)` will default to the item at index 30, not the one at index 100. 

Manually calculating offset to a specific index is not possible when the items are of variable in size because all the containers are not initialized at once.

**Reply by MikeHillberg** [reply]
Can you do this?

`(myListView.ContainerFromIndex(100) as UIElement)?.StartBringIntoView()`

**Reply by w-ahmad** [reply]
I tried this, but it returned `null`, possibly due to virtualization.

**Reply by ranjeshj** [reply]
I'm not sure if there is an easy way to do this with duplicate items since the APIs are based on object instead of Index. ItemsRepeater/ItemsView have APIs that are based on index, so that might be something you could see if it addresses your scenario.

**Reply by w-ahmad** [reply]
Thanks @ranjeshj

**Comment by amtopel** [comment]
Are there any plans to fix the XAML compiler bug [3331](https://github.com/microsoft/microsoft-ui-xaml/issues/3331), which breaks `winrt::implements` in C++/WinRT?  Thanks for your time and consideration.

**Comment by artkat** [comment]
WinUI still does not support the [x:TypeArguments Directive](https://learn.microsoft.com/en-us/dotnet/desktop/xaml-services/xtypearguments-directive) because it does not support XAML 2009. This makes it impossible to make generic controls such as MyUserControl.
When will WinUI support XAML 2009? MAUI and even WPF do support it.

**Comment by Tiktack** [comment]
Stable release for 1.6 planned on Fall 2024, but any estimated date when 1.6 preview will be available to public? Really wanna start playing with native AOT.

**Comment by eduardobragaxz** [comment]
Are in process background tasks and UWP-like lifecycle events planned?

**Reply by DarranRowe** [reply]
UWP like lifecycle events needs the support of Windows.
If you look at the [Application](https://learn.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-application) element of the package schema, there is desktop11:AppLifecycleBehaviour attribute. This seems to allow packaged desktop applications to have Windows send the lifecycle events. However, desktop11 isn't supported in 22621 or older.
The Windows 11 24H2 SDK does have the schema for desktop11, so I would guess that this currently needs a preview version of Windows.

**Comment by jschwizer99** [comment]
Are there any stability fixes in the 1.6 that could improve the multi-select and virtualization issues of the `treeview`? I couldn't find any hints in the release notes of the latest 1.6 preview. In our application, the `treeview` issues strongly affect stability. An excerpt of issues can be found in the March WinUI Community Call https://github.com/microsoft/microsoft-ui-xaml/discussions/9361#discussioncomment-8693039
At that time there was hope that some of these issues get addressed.

**Comment by HotCakeX** [comment]
Hello,
Can the WinUI team talk about their plans to make WinUI3 officially usable in PowerShell? PowerShell is based on .NET.

There are users on the PowerShell side who want to be able to use WinUI3 but there is no official guidance or even official example of how to do it.

I posted this question on StackOverflow and received some responses: https://stackoverflow.com/questions/78468949/how-to-create-winui3-gui-in-powershell
But none of them are perfect, not all WinUI3 features can be used in PowerShell through those methods. 99% of the time it's because we don't know exactly how to bootstrap WinUI3 in PowerShell, what components and DLLs are exactly required for all WinUI3 functionalities to work in PowerShell.

I'm hoping the WinUI team will go over this topic and provide some guidance, helping material and code examples for how to do this in PowerShell.

P.S I'm totally aware of WPF can be used in PowerShell and I've built GUI based on that and also aware that WPF is being overhauled at the end of this year with Windows 11 themes, however the focus of my post is only about WinUI3 in PowerShell.

When we use Visual Studio, it's extremely easy to setup a C# app that uses WinUI3 with pre-made setups, the same ease of use should be provided for PowerShell users as well or at least there should be a official Microsoft guide that explains how we can do it manually, so we can automate it ourselves through scripts and so on.

Thank you!
